### PR TITLE
Add warning if too many zero values for advection velocities

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses #571 

Description
Edits to optical flow to add a warning being raised, if an excess of zero values is calculated within the advection velocities. At the moment, the threshold for when a warning is raised is hard-coded within `optical_flow.py`. This could be modified, if this is desirable.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py [hunk 1]
```diff
@@ -473,6 +474,40 @@ def test_values(self):
         self.assertAlmostEqual(np.mean(vmat), 0.121514428331)
 
 
+class Test_zero_advection_velocities_warning(IrisTest):
+    """Test the zero_advection_velocities_warning."""
```
Comment (Katie-Howard) [inline lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py]: I wonder whether there should be another test here for when there is a zero in the array but it makes up less than 10% of numbers? That would just check that no warnings are raised in that instance?

File: lib/improver/nowcasting/optical_flow.py [hunk 2]
```diff
@@ -731,6 +733,40 @@ def calculate_displacement_vectors(self, partial_dx, partial_dy,
...
+        Args:
+            vel_comp (np.ndarray):
+                Advection velocity that will be checked to assess the
+                proportion of zeroes present in this field.
+
+        Keyword Args:
+            non_zero_vel_threshold (float):
+                Fractional value to specify the proportion of non-zero values
+                that the advection field should contain at a minimum.
+                For example, if non_zero_vel_threshold=0.9 then up to 90% of
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/nowcasting/optical_flow.py]: I've inverted this whole method, so that it's dealing with zero advection velocities, rather than non-zero advection velocities, as I think the focus is on the number of zero advection velocities exceeding the defined threshold.

File: lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py (hunk 1 above)
Comment (gavinevans) [reply lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py]: I've added a couple of unit tests. One, as you suggest, for where there are zeroes, but there are fewer zeroes than the threshold, and for where the threshold is modified.

File: lib/improver/nowcasting/optical_flow.py [hunk 3]
```diff
@@ -731,6 +733,40 @@ def calculate_displacement_vectors(self, partial_dx, partial_dy,
 
         return umat, vmat
 
+    @staticmethod
+    def zero_advection_velocities_warning(
+            vel_comp, zero_vel_threshold=0.1):
+        """
+        Raise warning if more than 10% of the cells within the domain have
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/nowcasting/optical_flow.py]: I've corrected this suggested.

**Comment by carolinejjones** [comment]
Rerun tests. Changes look good.
